Fix #1061: Story text scale with fragment manager

nikitamarysolomanpvt opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
duplicate of #1134

PTAL on the approach.
I had used different approach than that of textsize implementation done in options. I am scaling the actual size of the text rather than setting font size changes in a particular widget. 
Also introduced fragment manager for exploration fragment referring https://gist.github.com/BenHenning/0eb3e6139bcf2bd4bb790e323830b92f

      StoryTextSize.SMALL_TEXT_SIZE -> .8f
      StoryTextSize.MEDIUM_TEXT_SIZE -> 1.0f
      StoryTextSize.LARGE_TEXT_SIZE -> 1.2f
      StoryTextSize.EXTRA_LARGE_TEXT_SIZE -> 1.4f
Test cases are not included as am waiting for confirmation on the above.
(https://github.com/oppia/oppia-android/pull/1229#issuecomment-647969550)
      StoryTextSize.SMALL_TEXT_SIZE -> .8f
<img width="375" alt="small" src="https://user-images.githubusercontent.com/54615666/83108645-bb5ef300-a0dd-11ea-81c6-14a8d3cff0e4.png">
<img width="366" alt="small" src="https://user-images.githubusercontent.com/54615666/83108726-d7fb2b00-a0dd-11ea-8cd9-63e95f7f439a.png">

      StoryTextSize.MEDIUM_TEXT_SIZE -> 1.0f
<img width="373" alt="small" src="https://user-images.githubusercontent.com/54615666/83108984-37f1d180-a0de-11ea-8a05-2feee55f96de.png">
<img width="461" alt="small" src="https://user-images.githubusercontent.com/54615666/83109120-6e2f5100-a0de-11ea-821e-2c7dd5f4ee5e.png">

      StoryTextSize.LARGE_TEXT_SIZE -> 1.4f
<img width="461" alt="small" src="https://user-images.githubusercontent.com/54615666/83109249-b189bf80-a0de-11ea-8ec3-057a46b978ed.png">
<img width="461" alt="small" src="https://user-images.githubusercontent.com/54615666/83109360-e6961200-a0de-11ea-9320-bf490e381292.png">

      StoryTextSize.EXTRA_LARGE_TEXT_SIZE -> 1.6f

<img width="461" alt="small" src="https://user-images.githubusercontent.com/54615666/83109403-f9a8e200-a0de-11ea-8257-ecd2db854850.png">
<img width="461" alt="small" src="https://user-images.githubusercontent.com/54615666/83109422-03cae080-a0df-11ea-9970-997b89ad5443.png">

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

nikitamarysolomanpvt committed:
ic_onboarding_0 image update issue fix

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into develop

 Conflicts:
	app/src/main/res/layout-land/walkthrough_welcome_fragment.xml

nikitamarysolomanpvt committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into develop

nikitamarysolomanpvt committed:
created ExplorationManagerFragment for getting default text size for ExplorationFragment to update font scale.

nikitamarysolomanpvt committed:
KDoc and other nits.

nikitamarysolomanpvt committed:
nit

nikitamarysolomanpvt committed:
nit

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -21,9 +22,11 @@ private const val TAG_STOP_EXPLORATION_DIALOG = "STOP_EXPLORATION_DIALOG"
+  StateKeyboardButtonListener,
+  AudioButtonListener, HintsAndSolutionListener, RouteToHintsAndSolutionListener,
+  RevealHintListener,
+  RevealSolutionInterface, DeafultFontSizeStateListener {
```
To make it more readable follow one of the below approaches:
```
class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterface,
  StateKeyboardButtonListener, AudioButtonListener, HintsAndSolutionListener,
  RouteToHintsAndSolutionListener, RevealHintListener, RevealSolutionInterface,
  DeafultFontSizeStateListener {
```
class ExplorationActivity : InjectableAppCompatActivity(),
  StopExplorationInterface,
  StateKeyboardButtonListener,
  AudioButtonListener,
  HintsAndSolutionListener,
  RouteToHintsAndSolutionListener,
  RevealHintListener,
  RevealSolutionInterface,
  DeafultFontSizeStateListener {
```

```

nikitamarysolomanpvt replied:
Updated alignment for better readability.

veena14cs replied:
DeafultFontSizeStateListener can be in new line.

rt4914 replied:
+1

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -130,12 +145,20 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
     getHintsAndSolution()?.dismiss()
   }
+
+  override fun onDeafultFontSizeLoaded(result: StoryTextSize) {
```
Instead of `result` we can name it something more contextual `storyTextSize` or `defaultStoryTextSize`

nikitamarysolomanpvt replied:
updated to storyTextSize.

rt4914 replied:
You will need to make this change in `DeafultFontSizeStateListener` interface also.

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -13,15 +13,18 @@ import org.oppia.app.R
 import javax.inject.Inject
 
 const val TAG_EXPLORATION_FRAGMENT = "TAG_EXPLORATION_FRAGMENT"
+const val TAG_EXPLORATION_MANAGER_FRAGMENT = "TAG_EXPLORATION_MANAGER_FRAGMENT"
```
This should be private.

nikitamarysolomanpvt replied:
changed both constants to private.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -65,18 +79,46 @@ class ExplorationActivityPresenter @Inject constructor(
+      val explorationManagerFragment = ExplorationManagerFragment()
+      val args = Bundle()
+      args.putInt(
+        ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY,
```
Instead of this you should use key specific to this fragment.  Like:
`ExplorationManagerFragment.EXPLORATION_FRAGMENT_MANAGER_INTERNAL_PROFILE_ID_ARGUMENT_KEY`

Similar change in `ExplorationManagerFragment`

nikitamarysolomanpvt replied:
This is introduced in consistency with other keys that are used for sending values to fragment from ExplorationActivity PTAL on previous code.

rt4914 replied:
Keeping this open so that we can finalise this after team meeting.

My suggestion is this:
in `ExplorationActivityPresenter`
```
   if (getExplorationManagerFragment() == null) {
      activity.supportFragmentManager.beginTransaction().add(
        R.id.exploration_fragment_placeholder,
        ExplorationManagerFragment.newInstance(internalProfileId),
        ExplorationManagerFragment.TAG
      ).commitNow()
    }
```

In `ExplorationManagerFragment`
```
  companion object {
    internal const val EXPLORATION_MANAGER_FRAGMENT_INTERNAL_PROFILE_ID_ARGUMENT_KEY = "ExplorationManagerFragment.internal_profile_id"

const val TAG = "ExplorationManagerFragment.TAG"

    fun newInstance(internalProfileId: Int): ExplorationManagerFragment {
      val explorationManagerFragment = ExplorationManagerFragment()
      val args = Bundle()
      args.putInt(EXPLORATION_MANAGER_FRAGMENT_INTERNAL_PROFILE_ID_ARGUMENT_KEY, internalProfileId)
      explorationManagerFragment.arguments = args
      return explorationManagerFragment
    }
  }
```

nikitamarysolomanpvt replied:
Will discuss this in team meeting.

nikitamarysolomanpvt replied:
This will be included for entire exploration arguments in separate PR.

rt4914 replied:
@nikitamarysolomanpvt Please create a issue number accordingly and link it here.

nikitamarysolomanpvt replied:
Issue created https://github.com/oppia/oppia-android/issues/1267.

BenHenning replied:
Per the team meeting, is this something that can be resolved now?

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationFragment.kt:
```diff
@@ -14,6 +15,9 @@ class ExplorationFragment : InjectableFragment() {
     super.onAttach(context)
+    val storyTextSize =
+      arguments!!.getString(ExplorationActivity.EXPLORATION_ACTIVITY_STORY_DEFAULT_FONT_SIZE_ARGUMENT_KEY)
+    FontScaleConfigurationUtil(context, storyTextSize).adjustFontScale()
```
The `storyTextSize ` is a `String?` and the value which is passed in `FontScaleConfigurationUtil ` is `String`. Therefore make sure that the `storyTextSize` is non-null first.

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragment.kt:
```diff
@@ -0,0 +1,26 @@
+    val internalProfileId =
+      arguments!!.getInt(ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY, -1)
+    explorationManagerFragmentPresenter.handleCreate(internalProfileId)
+    setRetainInstance(true)
```
Nit: `retainInstance = true`

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragment.kt:
```diff
@@ -0,0 +1,26 @@
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    val internalProfileId =
+      arguments!!.getInt(ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY, -1)
```
Nit: `arguments!!.getInt(ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY, /* defaultValue= */ -1)`

nikitamarysolomanpvt replied:
updated as per ur comment.

rt4914 replied:
As this is taking multiple lines, comment should be inline of its value:
```
      arguments!!.getInt(
        ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY, 
        /* defaultValue= */ -1
      )
```

nikitamarysolomanpvt replied:
It's already updated in last PR...

rt4914 replied:
Actually the code is showing the inline comment in upper line and it should actually be in line of the value:

Before:
```
      arguments!!.getInt(
        ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY, /* defaultValue= */ 
        -1
      )
```

After:
```
      arguments!!.getInt(
        ExplorationActivity.EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY, 
        /* defaultValue= */ -1
      )
```

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragment.kt:
```diff
@@ -0,0 +1,26 @@
+import javax.inject.Inject
+
+/**
+ * FragmentManager of [ExplorationFragment] that observes data provider that retrive default story text size.
```
`FragmentManager` in android has a specific meaning and this class is not a `FragmentManager`.
So I think the kdoc can be updated to remove the use of that word.
Something like:
`Fragment manager of [ExplorationFragment] that observes data provider that retrieve default story text size.`

rt4914 replied:
Also typo: `retrieve`

nikitamarysolomanpvt replied:
updated to ManagerFragment and retrieve.

rt4914 commented on app/src/main/java/org/oppia/app/utility/FontScaleConfigurationUtil.kt:
```diff
@@ -0,0 +1,36 @@
+    val metrics: DisplayMetrics = context.getResources().getDisplayMetrics()
+    val windowManager =
+      context.getSystemService(WINDOW_SERVICE) as WindowManager?
+    windowManager!!.defaultDisplay.getMetrics(metrics)
```
Nit:
```
    val metrics: DisplayMetrics = context.resources.displayMetrics
    val windowManager = context.getSystemService(WINDOW_SERVICE) as WindowManager?
```

nikitamarysolomanpvt replied:
Updated.

rt4914 reviewed (changes requested):
Implementation is very clean. Suggested some nit changes.

nikitamarysolomanpvt committed:
nit

veena14cs commented on app/src/main/java/org/oppia/app/player/exploration/DeafultFontSizeStateListener.kt:
```diff
@@ -0,0 +1,8 @@
+
+/** Listener for default font size fetch. */
+interface DeafultFontSizeStateListener {
+  fun onDeafultFontSizeLoaded(result: StoryTextSize)
```
There is typo error. Please correct it to onDefault for both function and interface name.

nikitamarysolomanpvt replied:
Done

veena14cs commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -40,17 +43,29 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopExplorationInterf
+      topicId,
+      storyId,
+      explorationId,
+      backflowScreen
```
Rename it by following camel-casing `backFlowScreen`

rt4914 replied:
@veena14cs 
This is actually correct. Android studio is showing this incorrect but this is a valid word with the correct [meaning](https://www.google.com/search?q=backflow&oq=backflow&aqs=chrome.0.69i59j0l4j69i61l3.1879j0j7&sourceid=chrome&ie=UTF-8) that we want here.

rt4914 replied:
Its similar to `OnBoarding` and `Onboarding` where android studio suggests the first one but second is the correct one.

nikitamarysolomanpvt replied:
This is not part of this PR and also backflowScreen is correct as it can be used as a single word.

veena14cs replied:
Got it.

veena14cs commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -47,8 +51,18 @@ class ExplorationActivityPresenter @Inject constructor(
+    topicId: String,
+    storyId: String,
+    explorationId: String,
+    backflowScreen: Int?
```
Ditto.

nikitamarysolomanpvt replied:
ditto

rt4914 commented:
@nikitamarysolomanpvt I have either resolved or replied to all my comments mentioned earlier. PTAL. Thanks.

veena14cs reviewed (changes requested):
Suggested some changes.

nikitamarysolomanpvt commented:
> @nikitamarysolomanpvt I have either resolved or replied to all my comments mentioned earlier. PTAL. Thanks.

Updated PTAL

nikitamarysolomanpvt committed:
nit

veena14cs reviewed (approved):
LGTM.  Are you creating new PR for testcases.

nikitamarysolomanpvt commented:
> LGTM. Are you creating new PR for testcases.

need to get approval from ben on the approach then will create test-case.

rt4914 reviewed (changes requested):
There are only two unresolved comments left which can be finished after the team meeting. Also, please add test cases also.

nikitamarysolomanpvt committed:
nit

rt4914 reviewed (changes requested):
LGTM,
1. Please create an issue to track the implementation change of fragment new instance. This issue should be general and not specific to exploration.
2. Link that issue number in this PR.
3. Blocking this PR because of test cases, which will be done after Ben's approval on the approach.

nikitamarysolomanpvt commented:
> LGTM,
> 
> 1. Please create an issue to track the implementation change of fragment new instance. This issue should be general and not specific to exploration.
> 2. Link that issue number in this PR.

Done.

BenHenning commented on app/src/main/java/org/oppia/app/utility/FontScaleConfigurationUtil.kt:
```diff
@@ -0,0 +1,35 @@
+import org.oppia.app.model.StoryTextSize
+import javax.inject.Inject
+
+/** FontScaleConfigurationUtil helps to update font scale. */
```
Or perhaps: ``Utility to change the scale of font for the entire app.``

nikitamarysolomanpvt replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/app/utility/FontScaleConfigurationUtil.kt:
```diff
@@ -0,0 +1,35 @@
+import javax.inject.Inject
+
+/** FontScaleConfigurationUtil helps to update font scale. */
+class FontScaleConfigurationUtil @Inject constructor(
```
This class is being constructed, not innjected. We should inject it and pass the story text size in as a parameter.

nikitamarysolomanpvt replied:
Done. PTAL

BenHenning commented on app/src/main/java/org/oppia/app/utility/FontScaleConfigurationUtil.kt:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  /** This method returns font scale by story text size. */
+  private fun getFontScaleByStoryTextSize(): Float {
```
In retrospect, we should probably name this utility specifically to story text size since this is technically for that use case. Suggest:

``StoryTextSizeConfigurationUtil``

nikitamarysolomanpvt replied:
Renamed to getStoryTextSizeConfigurationUtil

BenHenning commented on app/src/main/java/org/oppia/app/utility/FontScaleConfigurationUtil.kt:
```diff
@@ -0,0 +1,35 @@
+    val configuration = context.resources.configuration
+    configuration.fontScale = getFontScaleByStoryTextSize()
+    val metrics: DisplayMetrics = context.resources.displayMetrics
+    val windowManager = context.getSystemService(WINDOW_SERVICE) as WindowManager?
```
Prefer: ``context.getSystemService(WINDOW_SERVICE) as? WindowManager``

This is actually a safer way to cast with nullability. See: https://kotlinlang.org/docs/reference/typecasts.html#safe-nullable-cast-operator.

nikitamarysolomanpvt replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragmentPresenter.kt:
```diff
@@ -0,0 +1,49 @@
+    })
+  }
+
+  private fun getProfileData(): LiveData<StoryTextSize> {
```
Is this method name correct? Perhaps instead: ``retrieveStoryTextSize`` would be clearer.

nikitamarysolomanpvt replied:
renamed to  retrieveStoryTextSize

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragmentPresenter.kt:
```diff
@@ -0,0 +1,49 @@
+
+  fun handleCreate(internalProfileId: Int) {
+    this.profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    getProfileData().observe(activity, Observer<StoryTextSize> { result ->
```
Prefer observing on the fragment since observers ought to correspond to the lifetime of the fragment, not the activity. Otherwise, this could partially leak references to the fragment.

nikitamarysolomanpvt replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragment.kt:
```diff
@@ -0,0 +1,30 @@
+        /* defaultValue= */ -1
+      )
+    explorationManagerFragmentPresenter.handleCreate(internalProfileId)
+    retainInstance = true
```
Why set this? Per https://developer.android.com/reference/androidx/fragment/app/Fragment#setRetainInstance(boolean) this also seems deprecated.

nikitamarysolomanpvt replied:
removed    ` retainInstance = true`

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -92,28 +134,41 @@ class ExplorationActivityPresenter @Inject constructor(
-          logger.d("ExplorationActivity", "Successfully stopped exploration")
-          backPressActivitySelector(backflowScreen)
-          (activity as ExplorationActivity).finish()
+    FontScaleConfigurationUtil(activity, StoryTextSize.MEDIUM_TEXT_SIZE.name).adjustFontScale()
```
Isn't this happening in the manager fragment, instead?

nikitamarysolomanpvt replied:
When i tried calling it in manager fragment it is updating 
UI late.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -92,28 +134,41 @@ class ExplorationActivityPresenter @Inject constructor(
+  private fun getExplorationManagerFragment(): ExplorationManagerFragment? {
+    return activity.supportFragmentManager.findFragmentByTag(
+      TAG_EXPLORATION_MANAGER_FRAGMENT
+    ) as ExplorationManagerFragment?
```
See later comment: please use safe-casting here, instead.

nikitamarysolomanpvt replied:
used safe-casting

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -65,18 +79,46 @@ class ExplorationActivityPresenter @Inject constructor(
+      )
       explorationFragment.arguments = args
-      activity.supportFragmentManager.beginTransaction().add(
+      activity.supportFragmentManager.beginTransaction().replace(
```
Why this change? We shouldn't replace since we'll lose the bundle the fragment uses to retain its state on configuration changes.

nikitamarysolomanpvt replied:
Removed

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt--I think the overall approach makes sense. I had some nits on it, but I didn't see anything major to comment on.

nikitamarysolomanpvt committed:
nit fixes and test activity

nikitamarysolomanpvt committed:
Merge branches 'develop' and 'story-text-scale-with-fragment-manager' of https://github.com/oppia/oppia-android into story-text-scale-with-fragment-manager

 Conflicts:
	app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt

nikitamarysolomanpvt committed:
change in arguments of ExplorationFragment and test-cases

nikitamarysolomanpvt commented:
> ## Explanation
> duplicate of #1134
> 
> PTAL on the approach.
> I had used different approach than that of textsize implementation done in options. I am scaling the actual size of the text rather than setting font size changes in a particular widget.
> Also introduced fragment manager for exploration fragment referring https://gist.github.com/BenHenning/0eb3e6139bcf2bd4bb790e323830b92f
> 
> ```
>   StoryTextSize.SMALL_TEXT_SIZE -> .8f
>   StoryTextSize.MEDIUM_TEXT_SIZE -> 1.0f
>   StoryTextSize.LARGE_TEXT_SIZE -> 1.2f
>   StoryTextSize.EXTRA_LARGE_TEXT_SIZE -> 1.4f
> ```
> 
> Test cases are not included as am waiting for confirmation on the above.

added test cases and also changed all arguments and arguments implementation in ExplorationFragment as per suggestions in meeting PTAL

veena14cs commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -40,17 +48,29 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopStatePlayingSessi
     internal const val EXPLORATION_ACTIVITY_STORY_ID_ARGUMENT_KEY = "ExplorationActivity.story_id"
-    internal const val EXPLORATION_ACTIVITY_EXPLORATION_ID_ARGUMENT_KEY = "ExplorationActivity.exploration_id"
-    internal const val EXPLORATION_ACTIVITY_BACKFLOW_SCREEN_KEY = "ExplorationActivity.backflow_screen"
+    internal const val EXPLORATION_ACTIVITY_STORY_DEFAULT_FONT_SIZE_ARGUMENT_KEY =
```
Please remove this if its not getting used anywhere.

nikitamarysolomanpvt replied:
Removed.

veena14cs commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivity.kt:
```diff
@@ -40,17 +48,29 @@ class ExplorationActivity : InjectableAppCompatActivity(), StopStatePlayingSessi
     /** Returns a new [Intent] to route to [ExplorationActivity] for a specified exploration. */
 
-    internal const val EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY = "ExplorationActivity.profile_id"
+    internal const val EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY =
```
Just to confirm, as per the discussion in the meeting regarding keys, is this suppose to be `EXPLORATION_ACTIVITY_PROFILE_ID_EXTRA_KEY` instead of `EXPLORATION_ACTIVITY_PROFILE_ID_ARGUMENT_KEY`. Ditto elsewhere

nikitamarysolomanpvt replied:
I updated only ExplorationFragment arguments as mentioned in the Explanation because this was not introduced by me. Will be tracked in #1267.

veena14cs replied:
Okay, got it.

veena14cs commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -148,16 +190,35 @@ class ExplorationActivityPresenter @Inject constructor(
     if (ephemeralStateResult.isFailure()) {
-      logger.e("StateFragment", "Failed to retrieve answer outcome", ephemeralStateResult.getErrorOrNull()!!)
+      logger.e(
+        "StateFragment",
```
Replace this log statement from `StateFragment` to `ExplorationActivity`.

nikitamarysolomanpvt replied:
Done.

veena14cs reviewed (approved):
LGTM beside some nit changes.

nikitamarysolomanpvt committed:
nit

rt4914 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -122,6 +122,9 @@
     <activity android:name=".testing.HomeTestActivity" />
     <activity android:name=".testing.InputInteractionViewTestActivity" />
+    <activity
+      android:name=".testing.FontScaleTestActivity"
```
This activity is not in alphabetical order.

nikitamarysolomanpvt replied:
Updated.

rt4914 commented on app/src/main/java/org/oppia/app/testing/FontScaleTestActivity.kt:
```diff
@@ -0,0 +1,29 @@
+
+private const val FONT_SCALE_EXTRA_KEY = "FONT_SCALE_EXTRA_KEY"
+
+/** Test Activity used for testing Font Scale */
```
Nit: `/** Test activity used for testing font scale. */`

nikitamarysolomanpvt replied:
updated as per suggestion.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/utility/FontSizeMatcher.kt:
```diff
@@ -0,0 +1,43 @@
+
+  companion object {
+    /**
+     * Matches that the check Font size.
```
Nit: `Matcher that the checks the font size.`

nikitamarysolomanpvt replied:
Updated to Matcher that checks the font size.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/utility/FontSizeMatcher.kt:
```diff
@@ -0,0 +1,43 @@
+
+// https://stackoverflow.com/questions/47398659/check-the-font-size-height-and-width-of-the-edittext-with-espresso
+/** This class mainly provides a custom matcher to test whether the text matches expected font size. */
+class FontSizeMatcher(expectedSize: Float) :
```
Nit update: `class FontSizeMatcher(private val expectedSize: Float) :` and after this you can remove the `private val expectedSize: Float`

nikitamarysolomanpvt replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleTestActivityTest.kt:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  @Test
+  fun testFontScaleTestActivity_smallTextSizeIsVerifiedSucessfully() {
```
Typo in the word `Sucessfully` -> `Successfully`. Here and everywhere else.

nikitamarysolomanpvt replied:
updated all method name typo errors

rt4914 reviewed (approved):
LGTM, suggested some nit changes.

nikitamarysolomanpvt committed:
nit

rt4914 commented on app/src/sharedTest/java/org/oppia/app/utility/FontSizeMatcher.kt:
```diff
@@ -0,0 +1,38 @@
+
+  companion object {
+    /**
+     * Matches that check font size.
```
This is still incorrect. (Instead of `Matches` use `Matcher`)
`Matcher that checks font size.`

nikitamarysolomanpvt replied:
Done

rt4914 reviewed (approved):
LGTM, just one final nit which needs to be fixed.

veena14cs reviewed (approved):
LGTM.

nikitamarysolomanpvt committed:
nit

rt4914 reviewed (approved):
LGTM, thanks.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationManagerFragmentPresenter.kt:
```diff
@@ -18,18 +19,19 @@ import javax.inject.Inject
     this.profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
-    getProfileData().observe(activity, Observer<StoryTextSize> { result ->
+    retrieveStoryTextSize().observe(fragment, Observer<StoryTextSize> { result ->
       (activity as DefaultFontSizeStateListener).onDefaultFontSizeLoaded(result)
```
I was actually expecting that the manager fragment would just go ahead and resize the text itself, rather than blocking loading the exploration fragment. Shouldn't we eventually be able to resize text size from within the player? Resizing the font scale here simplifies the solution a bit, and should automatically work with that later feature once we finish hooking up the overflow menu part of it.

nikitamarysolomanpvt replied:
@BenHenning Adding conversation context
> Here's what I was thinking: https://gist.github.com/BenHenning/0eb3e6139bcf2bd4bb790e323830b92f. The idea 
> is to have a fragment whose job is to just observe a data provider & then enable the activity to add/remove 
> fragments` or otherwise change its UI based on those results. I'm not sure if I explained this well--does it make sense to you?

@BenHenning We need to do resize within onAttach, Hence manager fragment just observe a data provider & then enable the activity to add fragment and helps to update in the onAttach of ExplorationFragment.

nikitamarysolomanpvt replied:
@BenHenning  If I'm scaling after retrieveStoryTextSize then as initially informed scaling is not working as expected, for eg: on configuration change scale changes, if user change scale more than three times it won't work and so on.

BenHenning commented on app/src/main/java/org/oppia/app/testing/FontScaleTestActivity.kt:
```diff
@@ -0,0 +1,29 @@
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+    val storyTextSize = intent.getStringExtra(FONT_SCALE_EXTRA_KEY)
+    FontScaleConfigurationUtil().adjustFontScale(this, storyTextSize)
```
Not keen on this--can we inject it as a field, instead?

nikitamarysolomanpvt replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/app/utility/FontScaleConfigurationUtil.kt:
```diff
@@ -7,29 +7,29 @@ import android.view.WindowManager
-      StoryTextSize.LARGE_TEXT_SIZE.name -> 1.4f
-      else -> 1.6f
+      StoryTextSize.LARGE_TEXT_SIZE.name -> 1.2f
+      else -> 1.4f
```
Is this supposed to be EXTRA_LARGE? Why default to the largest value? If we're missing any constants, I suggest adding all constants here and keeping the 'else' case as a reasonable default (like 1.0f).

nikitamarysolomanpvt replied:
Updated. to 

```     
      StoryTextSize.SMALL_TEXT_SIZE.name -> .8f
      StoryTextSize.MEDIUM_TEXT_SIZE.name -> 1.0f
      StoryTextSize.LARGE_TEXT_SIZE.name -> 1.2f
      StoryTextSize.EXTRA_LARGE_TEXT_SIZE.name -> 1.4f
      else -> 1.0f

BenHenning commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleTestActivityTest.kt:
```diff
@@ -0,0 +1,93 @@
+@RunWith(AndroidJUnit4::class)
+class FontScaleTestActivityTest {
+  lateinit var context: Context
+  @Before
```
Newline before this per https://developer.android.com/kotlin/style-guide?authuser=3#vertical.

nikitamarysolomanpvt replied:
updated.

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/FontSizeMatcher.kt:
```diff
@@ -0,0 +1,38 @@
+    if (target !is TextView) {
+      return false
+    }
+    return target.textSize == expectedSize
```
I suggest using our approximately equals check here since exact float equality is likely to fail for computed values (due to floating point rounding error).

nikitamarysolomanpvt replied:
Updated to `return target.textSize.equals(expectedSize)`

BenHenning commented on app/src/sharedTest/java/org/oppia/app/utility/FontSizeMatcher.kt:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  override fun describeTo(description: Description) {
+    description.appendText("with fontSize: ")
```
Suggested changes (including a small rephrasing):

```kotlin
description.appendText("with font size:").appendValue(expectedSize)
```

nikitamarysolomanpvt replied:
updated to `description.appendText("with font size:").appendValue(expectedSize)`

BenHenning commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleTestActivityTest.kt:
```diff
@@ -0,0 +1,93 @@
+
+/** Tests for [FontScaleTestActivity]. */
+@RunWith(AndroidJUnit4::class)
+class FontScaleTestActivityTest {
```
Please also add tests to verify that the exploration activity properly resizes font when the user enters & leaves it when the user sets their their font scale. This test suite is great for verifying the correctness of FontScaleConfigurationUtil, but doesn't verify that the exploration font size is correctly resized.

nikitamarysolomanpvt replied:
@BenHenning can we implement this after finalising the implementation of ExplorationManagerFragment?

BenHenning commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleTestActivityTest.kt:
```diff
@@ -0,0 +1,93 @@
+
+/** Tests for [FontScaleTestActivity]. */
+@RunWith(AndroidJUnit4::class)
+class FontScaleTestActivityTest {
```
I suggest FontScaleConfigurationUtilActivityTest as an alternative name since this is specifically testing the utility.

nikitamarysolomanpvt replied:
updated to FontScaleConfigurationUtilActivityTest.

BenHenning reviewed (changes requested):
Thanks @nikitamarysolomanpvt! Overall, I think the solution looks good. I have some thoughts on how we might be able to simplify the solution by avoiding the listener, left some nits, and think we should add exploration-specific tests.

nikitamarysolomanpvt committed:
renamed FontScaleTestActivity to FontScaleConfigurationUtilActivity and other nit changes.

nikitamarysolomanpvt committed:
nit

BenHenning commented:
@rt4914 can you please bring this up-to-date with develop?

Commit:
Resolve merge conflicts

Commit:
List issues fixes

Commit:
Added max length comment

BenHenning commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleConfigurationUtilActivityTest.kt:
```diff
@@ -0,0 +1,102 @@
+    ).use {
+      onView(withId(R.id.font_scale_content_text_view)).check(
+        matches(
+          FontSizeMatcher.withFontSize(
```
Suggest importing ``withFontSize`` directly to help with readability & compactness here & elsewhere in this test suite.

rt4914 replied:
Done on behalf of @nikitamarysolomanpvt

BenHenning commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleConfigurationUtilActivityTest.kt:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  @Test
+  fun testFontScaleTestActivity_smallTextSizeIsVerifiedSuccessfully() {
```
``testFontScaleConfigurationUtil_smallTextSize_hasCorrectDimension``

Some clarifications:
- We're testing the configuration util, not the test activity (the test activity is just a medium for testing the configuration util)
- It's important to be clear what the verification is so that authors can verify it as correct

Ditto for other tests below.

rt4914 replied:
Done on behalf of @nikitamarysolomanpvt for all test cases

BenHenning commented on app/src/main/java/org/oppia/app/testing/FontScaleConfigurationUtilActivity.kt:
```diff
@@ -0,0 +1,32 @@
+import javax.inject.Inject
+
+/** Test activity used for testing font scale. */
+class FontScaleConfigurationUtilActivity : InjectableAppCompatActivity() {
```
Prefer naming this ``TestFontScaleConfigurationUtilActivity`` & ditto for its presenter (just to be clear that this is test-only).

rt4914 replied:
Done for this file and presenter.

BenHenning commented on app/src/main/java/org/oppia/app/testing/FontScaleConfigurationUtilActivity.kt:
```diff
@@ -0,0 +1,32 @@
+class FontScaleConfigurationUtilActivity : InjectableAppCompatActivity() {
+
+  @Inject
+  lateinit var fontScaleConfigurationUtilActivityPresenter: FontScaleConfigurationUtilActivityPresenter // ktlint-disable max-line-length
```
Suggest renaming variables to avoid the line limit issue, e.g.

```kotlin
@Inject
lateinit var configUtilActivityPresenter: FontScaleConfigurationUtilActivityPresenter
```

rt4914 replied:
Updated to 
```
  @Inject
  lateinit var configUtilActivityPresenter: TestFontScaleConfigurationUtilActivityPresenter
```

BenHenning commented on app/src/sharedTest/java/org/oppia/app/testing/FontScaleConfigurationUtilActivityTest.kt:
```diff
@@ -0,0 +1,102 @@
+
+/** Tests for [FontScaleConfigurationUtilActivity]. */
+@RunWith(AndroidJUnit4::class)
+class FontScaleConfigurationUtilActivityTest {
```
Nice tests.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/DefaultFontSizeStateListener.kt:
```diff
@@ -0,0 +1,8 @@
+
+/** Listener for default font size fetch. */
+interface DefaultFontSizeStateListener {
+  fun onDefaultFontSizeLoaded(storyTextSize: StoryTextSize)
```
Please document this with specifics of when in the lifecycle it will be called, and whether it's safe to perform fragment transactions in the callback (this is something we should deterministically make sure of one way or another).

rt4914 replied:
Done.

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/DefaultFontSizeStateListener.kt:
```diff
@@ -0,0 +1,8 @@
+
+import org.oppia.app.model.StoryTextSize
+
+/** Listener for default font size fetch. */
```
Please add details about why this listener is necessary (e.g. to capture the context you explained to me about the need to set things up before the fragment is attached).

rt4914 replied:
Added a detailed comment explaining how we are using the headless fragment before loading the `ExplorationFragment`

```
/**
 * To set the font-size correctly we need to fetch it before loading the fragment and
 * therefore this listens to the default font size in `ExplorationManagerFragment` and passes
 * the information to `ExplorationActivity` which eventually loads the `ExplorationFragment`.
 */
```

BenHenning commented on app/src/main/java/org/oppia/app/player/exploration/ExplorationActivityPresenter.kt:
```diff
@@ -76,26 +81,36 @@ class ExplorationActivityPresenter @Inject constructor(
+    }
+  }
+
+  fun loadExplorationFragment(result: StoryTextSize) {
```
Nit: s/``result``/``storyTextSize``/ for consistency. It's not necessarily a 'result' here.

rt4914 replied:
Done.

BenHenning reviewed (approved):
Thanks @nikitamarysolomanpvt! The PR generally LGTM, I just had some nits that I think need to be addressed.

BenHenning committed:
Merge branch 'develop' into story-text-scale-with-fragment-manager

Commit:
Nit changes on behalf of Nikita

Commit:
Merge with develop

Commit:
Nit fix
